src/profile-logic/process-profile.js

```diff
@@ -1052,56 +1056,195 @@ export function processProfile(
+ * information that user doesn't want to share.
  */
 export function serializeProfile(
   profile: Profile,
-  includeNetworkUrls: boolean = true
+  PIIToBeRemoved?: PII
 ): string {
   // stringTable -> stringArray
   let urlCounter = 0;
   const newProfile = Object.assign({}, profile, {
-    meta: { ...profile.meta, networkURLsRemoved: !includeNetworkUrls },
+    meta: {
+      ...profile.meta,
+      networkURLsRemoved:
+        PIIToBeRemoved !== undefined && PIIToBeRemoved.networkUrls,
+      extensions:
+        PIIToBeRemoved !== undefined && PIIToBeRemoved.extensions
+          ? getEmptyExtensions()
+          : profile.meta.extensions,
+    },
     pages:
-      includeNetworkUrls === false && profile.pages
+      PIIToBeRemoved !== undefined &&
+      PIIToBeRemoved.networkUrls &&
+      profile.pages
         ? profile.pages.map(page =>
             Object.assign({}, page, {
               url: 'Page #' + urlCounter++,
             })
           )
         : profile.pages,
-    threads: profile.threads.map(thread => {
-      const stringArray = thread.stringTable.serializeToArray();
-      const newThread = Object.assign({}, thread);
-      delete newThread.stringTable;
-      if (includeNetworkUrls === false) {
-        for (let i = 0; i < newThread.markers.length; i++) {
-          const currentMarker = newThread.markers.data[i];
-          if (
-            currentMarker &&
-            currentMarker.type &&
-            currentMarker.type === 'Network'
-          ) {
-            // Remove the URI fields from marker payload.
-            currentMarker.URI = '';
-            currentMarker.RedirectURI = '';
-            // Strip the URL from the marker name
-            const stringIndex = newThread.markers.name[i];
-            stringArray[stringIndex] = stringArray[stringIndex].replace(
-              /:.*/,
-              ''
-            );
-          }
+    threads: profile.threads
+      .map((thread, threadIndex) => {
+        const stringArray = thread.stringTable.serializeToArray();
+        let newThread = Object.assign({}, thread);
+        delete newThread.stringTable;
+
+        if (PIIToBeRemoved !== undefined) {
+          newThread = removeThreadPII(
+            newThread,
+            threadIndex,
+            stringArray,
+            PIIToBeRemoved
+          );
         }
-      }
-      newThread.stringArray = stringArray;
-      return newThread;
-    }),
+
+        if (newThread === null) {
+          return null;
+        }
+
+        newThread.stringArray = stringArray;
+        return newThread;
+      })
+      .filter(thread => thread !== null),
   });
+
   return JSON.stringify(newProfile);
 }
 
+/**
+ * Take a thread with PII that user wants to be removed and remove the thread
+ * data depending on that PII status. Look at the `PII` type definition if you
+ * want to learn what kind of information we are removing.
+ */
+function removeThreadPII(
```

Comment:
This function successively removes various things, maybe it could be split in several functions to make them smaller and easier to read?